### ISSUE TITLE
Duplicate boxes when using `brick` + `vty-windows`

### ISSUE THREAD

**Issue by RyanGlScott** [chhackett/vty-windows#10]
I recently tried running the [`brick-dialog-demo`](https://github.com/jtdaugherty/brick/blob/3852513d2136005ac610c490246f34284592adcc/programs/DialogDemo.hs) program (from the [`brick`](https://github.com/jtdaugherty/brick) repo) on Windows, which is now possible thanks to `vty-windows`. When I first start the program by running this command from the `brick` repo checkout:

```
$ cabal run -fdemos brick-dialog-demo
```

Everything looks fairly reasonable to begin with:

![brick-windows-1](https://github.com/chhackett/vty-windows/assets/2364661/513f2d48-4092-418f-8d16-cce9836411e7)

(I do see some strange question mark symbols on each line, which might be related to Unicode/escape characters that my font doesn't support. In any case, that's not what this issue is about.)

If I hit the Right Arrow key to change my selection from "Red" to "Blue", however, then I encounter a graphical oddity:

![brick-windows-2](https://github.com/chhackett/vty-windows/assets/2364661/c88bb4f0-688e-48a6-a557-3e2d6a601da4)

The line that contains the color options is now duplicated! The bottom line of color options will change depending on which arrow keys I hit, but the top line of color options will remain unchanged. This doesn't happen when I run the program on Linux, on the other hand, so I think this indicates a bug in `vty-windows` (or perhaps some misconfiguration on my end). I observe this issue in PowerShell, `cmd.exe`, and the MSYS2 terminal.

I filed this issue under the `vty-windows` repo rather than the `brick` repo, as I suspect the issue is Windows-specific. This is just a guess, however, so let me know if this issue belongs elsewhere.

**Comment by jtdaugherty** [comment]
FWIW, I think the artifact in the image (which is at the start of most of the lines on the screen as well as after each dialog box line) may be related to attribute resetting. `vty` emits an attribute reset after each sequence of characters that had attributes applied to it, where a reset to the background color is needed.

**Comment by jtdaugherty** [comment]
Another possible data point: @RyanGlScott and I discussed this when Ryan tried a more sophisticated example program and it looks like in that case the input was being echoed to the terminal, which is what resulted in layout corruption.

**Comment by jtdaugherty** [comment]
I also found `vty-crossplatform`'s `vty-mode-demo` to emit the same artifacts in PowerShell as shown in the screenshots above. So that part, at least, isn't `brick`-specific.

**Comment by jtdaugherty** [comment]
Here's a screenshot.
![Screenshot 2023-11-01 182525](https://github.com/chhackett/vty-windows/assets/427221/98635f53-845d-45ff-a46c-99206832731c)

**Comment by jtdaugherty** [comment]
Also, the strange behavior Ryan showed above can be triggered by other demos. For example, `brick-list-demo` looks fine on initial rendering, but a single input event (such as `Down`) triggers a redraw that gets corrupted for some reason.

**Comment by jtdaugherty** [comment]
I wonder whether the artifact characters are actually responsible for the layout corruption. If those get fixed first, I wonder if the other issue just goes away.

**Comment by ShrykeWindgrace** [comment]
@RyanGlScott thanks for reporting that.
I suspect that `conhost` version is not up to the task, so a couple of questions:
- what is your windows version? (e.g. output of `winver`)
- how do you run your shells? By running them directly as executables or spawning from `Windows Terminal`?
- what is your `conhost` version (`get-command conhost` in a `pwsh` session should show that)

These artifact characters do not seem to be from unicode misconfiguration, but better be safe than sorry, what is your default codepage (e.g. `chcp` in a `pwsh` session) and font?

That being said, I do not have access to a windows build environment till next week, so I won't be able to reproduce your problem locally =(

**Comment by chhackett** [comment]
Thanks for the report @RyanGlScott.
I was able to reproduce it on my system after uninstalling Windows Terminal. Apparently Windows Terminal hid this behavior from me. I'll see what I can do in the next few days to address this.

In the meantime, you can obtain best results by installing and using Windows Terminal...

**Comment by jtdaugherty** [comment]
FWIW, my testing was in PowerShell in Windows 10. (As a developer on Windows I have no idea what I'm doing, and that's the first thing I tried to run. I'm not even sure I know how to run anything else terminal-like.) I installed `ghcup`, then installed `ghc` and `cabal`, cloned `brick` and `vty-crossplatform`, tried a build, and then ran some demos.

**Comment by RyanGlScott** [comment]
In case it's helpful:

* Microsoft Windows 10 Home, Version 10.0.19045 Build 19045
* I am running PowerShell/`cmd.exe` directly (i.e., not through Windows Terminal)
* My `conhost` version is:

  ```
  $ Get-Command conhost
  
  CommandType     Name                                               Version    Source
  -----------     ----                                               -------    ------
  Application     conhost.exe                                        10.0.19... C:\Windows\system32\conhost.exe
  ```
* My code page (as reported by `chcp`) is 437. I tried changing the code page to 65001, but the issues persisted.
* My font is Consolas (the default PowerShell/`cmd.exe` font, as far as I am aware).

**Comment by chhackett** [comment]
@jtdaugherty  - your guess was correct. The attribute reset sequence was incorrect, and somehow the terminal I was testing with accepted the incorrect sequence.

I've fixed that problem in version 0.1.0.2.

@RyanGlScott  - please let me know if you see other issues. Thank you!

**Comment by jtdaugherty** [comment]
Awesome! I tried that release in another `matterhorn` build and the UI looks *much* better. There are still a couple of artifacts left, so I'll dig into those and update here when I have more info.

**Comment by chhackett** [comment]
So I am suspecting that these capabilities are incorrect:
("setab", csi "%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m")
("setaf", csi "%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m")
("sgr", csi "0%?%p6%t;1%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p5%t;2%;%?%p7%t;8%;m%?%p9%t\SO%e\SI%;")

**Comment by jtdaugherty** [comment]
I'm not sure what those correspond to off hand, but! I narrowed things down to `reverseVideo` maybe being broken in `vty-windows`.

**Comment by jtdaugherty** [comment]
(That's what the remaining artifacts I'm seeing correspond to: in particular, the termination of `reverseVideo`. The reversed text itself looks as I would expect.)

**Comment by jtdaugherty** [comment]
Although now I'm having no success reproducing it in a modified version of `vty-crossplatform`'s `vty-event-echo` program.

**Comment by jtdaugherty** [comment]
Here's a reproduction of the issue using `brick`'s `brick-viewport-scrollbars-demo` with `reverseVideo` applied to the scroll bar handles.
![Screenshot 2023-11-02 192736](https://github.com/chhackett/vty-windows/assets/427221/b1fa49b1-a539-48a3-aaf7-0b2a335dd70a)

**Comment by jtdaugherty** [comment]
So it doesn't happen in every case; perhaps it's only for certain attribute transitions..

**Comment by jtdaugherty** [comment]
Here is a more minimal reproducer using `vty-crossplatform` only:
```haskell
{-# LANGUAGE FlexibleContexts #-}
module Main where

import qualified Graphics.Vty as V
import Graphics.Vty ((<->), (<|>))
import Graphics.Vty.CrossPlatform (mkVty)

import Control.Arrow
import Control.Monad (unless)
import Control.Monad.IO.Class (MonadIO(..))
import Control.Monad.RWS (MonadReader(..), MonadState(..), RWST, execRWST, modify)

import Data.Sequence (Seq, (<|) )
import qualified Data.Sequence as Seq

eventBufferSize :: Int
eventBufferSize = 1000

type App = RWST V.Vty () (Seq String) IO

main :: IO ()
main = do
    vty <- mkVty V.defaultConfig
    _ <- execRWST (vtyInteract False) vty Seq.empty
    V.shutdown vty

vtyInteract :: Bool -> App ()
vtyInteract shouldExit = do
    updateDisplay
    unless shouldExit $ handleNextEvent >>= vtyInteract

updateDisplay :: App ()
updateDisplay = do
    let info = V.string (V.defAttr `V.withStyle` V.reverseVideo) "Press ESC to exit."
        extra = V.string (V.defAttr `V.withForeColor` V.red) "Red"
    eventLog <- foldMap (V.string V.defAttr) <$> get
    let pic = V.picForImage $ (info <|> extra) <-> eventLog
    vty <- ask
    liftIO $ V.update vty pic

handleNextEvent :: App Bool
handleNextEvent = ask >>= liftIO . V.nextEvent >>= handleEvent
    where
        handleEvent e = do
            modify $ (<|) (show e) >>> Seq.take eventBufferSize
            return $ e == V.EvKey V.KEsc []
```
Screenshot:
![Screenshot 2023-11-02 193721](https://github.com/chhackett/vty-windows/assets/427221/a75398ff-955b-43be-aa57-1e4bdca4ef6c)

**Comment by jtdaugherty** [comment]
That's `vty-event-echo` but with a transition from `reverseVideo` to red text.

**Comment by jtdaugherty** [comment]
Along with this I discovered two other issues:

* Using `bold` instead of `reverseVideo` in the demo above results in the same artifact.
* Using `bold` at all gives yellow text, rather than bolded text of whatever color it would have otherwise had.

**Comment by chhackett** [comment]
@jtdaugherty 
So I believe I've fixed the remaining invalid character issues (I cleaned up the sgr capability sequence), but I was wondering if you'd be willing to try my changes with your matterhorn app?
My changes are checked in to the [fix-bad-output](https://github.com/chhackett/vty-windows/tree/fix-bad-output) branch.

Thanks!

**Comment by RyanGlScott** [comment]
I just running `matterhorn` using `vty-windows`' `fix-bad-output` branch, and it works _beautifully_. No invalid characters, no duplicate lines—everything pretty much Just Works™. Perhaps @jtdaugherty would want to give it a try to see if everything is up to his standards, but as far as I am concerned, that branch resolves all of the issues I was having.

**Comment by chhackett** [comment]
Thanks for testing it, @RyanGlScott. I'll get this change merged in today.

**Comment by jtdaugherty** [comment]
I can confirm that:

* `matterhorn` is runs better with no artifacts that I can see
* the modified demo program runs with no artifacts

I did notice that the bold text is still yellow in PowerShell, but I don't know whether that's just a PowerShell palette/font configuration thing. I also noticed that Matterhorn's `C-g` channel selection mode behaves a little strangely in that the attribute it uses for selection match text (black foreground, magenta background, and underline) shows up without the magenta background color. But I think we can save that for another ticket.

Thanks for your fixes, @chhackett!
